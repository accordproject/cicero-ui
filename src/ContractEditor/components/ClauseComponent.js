import React from 'react';
import PropTypes from 'prop-types';

import { Icon } from 'semantic-ui-react';
import styled from 'styled-components';

import BaseComponent from './BaseComponent';
import TextBlock from './TextBlock';

const ClauseWrapper = styled(TextBlock)`
  margin: 10px auto;
  padding: 0;
  border: 1px solid #95D0FF;
  background-color: #E3F6FF;
  box-shadow: 0 1px 9px 0 rgba(0,124,232,0.16);
`;

const ClauseErrorWrapper = styled(ClauseWrapper)`
  border: 1px solid #FF9595;
  background-color: #FFEAE3;
  box-shadow: 0 1px 9px 0 rgba(232,0,0,0.16);
`;

const ClauseMenu = styled.div`
  position: relative;
  margin: 0 0 10px;
  padding: 10px 20px 8px;
  border-bottom: 1px solid #95D0FF;
`;

const ContractActions = styled.div`
  float: right;
  padding-top: 2px;
`;

const ClauseAction = styled.a`
  color: #99D2FB;
  cursor: pointer;
`;

const MenuText = styled.p`
  margin-bottom: 2px;
  color: #2587DA;
  font-size: 12px;
  line-height: 15px;
`;

const ClauseName = styled(MenuText)`
  font-weight: bold;
`;

const ClauseText = styled.p`
  margin: 0;
  padding: 0 20px 10px;
`;

const ClauseError = styled.pre`
  position: relative;
  margin: 0;
  padding: 10px 20px 8px;
  border-top: 1px solid #95D0FF;
  overflow-x: auto;
`;

class ClauseComponent extends BaseComponent {
  constructor(props) {
    super(props);
    this.handleDelete = this.handleDelete.bind(this);
    this.handleEdit = this.handleEdit.bind(this);
  }

  getClause() {
    return this.props.node.data.get('clause');
  }

  handleDelete() {
    this.props.editor.props.onClauseEvent(
      'handleDeleteClause',
      { id: this.props.node.data.get('clauseId') },
    );
  }

  handleEdit() {
    this.props.editor.props.onClauseEvent(
      'handleEditClause',
      { id: this.props.node.data.get('clauseId') },
    );
  }

  renderMenu() {
    const { node, editor } = this.props;
    const readOnly = editor.props.readOnly;
    return (
      <ClauseMenu contentEditable={false}>
        { readOnly ? null : (
          <ContractActions>
            <ClauseAction onClick={this.handleEdit}>
              <Icon name="pencil" size="large" />
            </ClauseAction>
            <ClauseAction onClick={this.handleDelete}>
              <Icon name="trash" size="large" />
            </ClauseAction>
          </ContractActions>
        )}
        <ClauseName className="templateName">
          {node.data.get('templateName')}
        </ClauseName>
        <MenuText>
          Only certain parts of the Smart Clause text can be edited
        </MenuText>
      </ClauseMenu>
    );
  }

  render() {
    const { node, children } = this.props;
    let Wrapper = ClauseWrapper;
    let err = null;
    if (!node.data.get('valid')) {
      Wrapper = ClauseErrorWrapper;
      err = node.data.get('error');
    }

    return (
      <Wrapper templateuri={node.data.get('templateUri')} {...this.props.attributes}>
        {this.renderMenu()}
        <ClauseText>
          {children}
        </ClauseText>
        { !err ? null : (
          <ClauseError contentEditable={false}>{err}</ClauseError>
        )}
      </Wrapper>
    );
  }
}

ClauseComponent.propTypes = {
  children: PropTypes.arrayOf(PropTypes.object).isRequired,
  editor: PropTypes.object.isRequired,
  node: PropTypes.object.isRequired,
};

export default ClauseComponent;
