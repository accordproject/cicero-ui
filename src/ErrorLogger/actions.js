export const gtZero = input => input > 0;

export const typeSwitchCase = (input) => {
  switch (true) {
    case ('modelError' in input):
      return 'Model';
    case ('parseError' in input):
      return 'Grammar';
    default:
      return 'Unknown';
  }
};

export const keySwitchCase = (input) => {
  switch (true) {
    case ('modelError' in input):
      return input.clauseTemplateId;
    case ('parseError' in input):
      return input.clauseId;
    default:
      return null;
  }
};

export const overalltypeSwitchCase = (input) => {
  switch (true) {
    case ('modelError' in input):
      return input.modelError;
    case ('parseError' in input):
      return input.parseError;
    default:
      return null;
  }
};

export const errorsExist = errors => gtZero(errors.length);

export const errorArrayLength = errors => (errorsExist(errors) ? errors.length : 'No');

export const isMultpleErrors = errors => ((errors.length > 1) ? 'Errors' : 'Error');

export const truncateMessage = string => ((string.length > 200)
  ? `${string.substring(0, 200)}...` : string);
