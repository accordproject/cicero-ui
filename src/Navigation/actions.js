const isClause = input => input === 'clause';
const isHeadingOne = input => input === 'heading_one';
const isHeadingTwo = input => input === 'heading_two';
const isHeadingThree = input => input === 'heading_three';

const truncateHeading = (heading, length) => ((heading.length > length)
  ? `${heading.substring(0, length)}...` : heading);


export const truncateHeader = ({ type, text }) => {
  if (isClause(type)) return truncateHeading(text, 22);
  if (isHeadingOne(type)) return truncateHeading(text, 22);
  if (isHeadingTwo(type)) return truncateHeading(text, 18);
  if (isHeadingThree(type)) return truncateHeading(text, 14);
  return 'Error!';
};
