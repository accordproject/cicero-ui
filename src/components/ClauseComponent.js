/* React */
import React, { useState, useEffect } from 'react';
// import { Inline, Block } from 'slate';
import PropTypes from 'prop-types';

/* Styling */
import * as S from './styles';

/* Icons */
import * as deleteIcon from '../icons/trash';
import * as editIcon from '../icons/edit';
import * as testIcon from '../icons/testIcon';

/* Actions */
import {
  headerGenerator,
  titleGenerator
} from './actions';

/* Components */
// import ConditionalBoolean from './ConditionalBoolean';
// import ClauseVariableList from './ClauseVariableList';

/* slate-command */
// import regenerateKey from '../SlateCommands/RegenerateAllNodeKeys';

/**
 * Component to render a clause
 * This will have an id property of the clauseid
 * @param {*} props
 */
function ClauseComponent(props) {
  const clauseProps = props.clauseProps || Object.create(null);

  // Tooltip visibility controls
  const [hovering, setHovering] = useState(false);
  const [hoveringHeader, setHoveringHeader] = useState(false);
  const [hoveringTestIcon, setHoveringTestIcon] = useState(false);
  const [hoveringEditIcon, setHoveringEditIcon] = useState(false);
  const [hoveringDeleteIcon, setHoveringDeleteIcon] = useState(false);

  // const [conditionals, setConditionals] = useState({});
  // const [listVariables, setListVariables] = useState({});

  const title = titleGenerator(props.templateUri);
  const header = headerGenerator(props.templateUri, clauseProps.HEADER_TITLE);

  const iconWrapperProps = {
    currentHover: hovering,
    // iconBg: clauseProps.CLAUSE_BACKGROUND
  };

  /* eslint react/no-find-dom-node: "off" */
  // useEffect(() => {
  //   const findPosition = (key) => {
  //     const pathToConditional = props.editor.value.document.getPath(key);
  //     const elementDOM = props.editor.findDOMNode(pathToConditional);
  //     const popupPosition = 'bottom left';
  //     const positionalStyle = {
  //       popupStyle: { top: elementDOM.offsetTop, left: elementDOM.offsetLeft, transform: 'none' },
  //       popupPosition,
  //       popupHeight: elementDOM.offsetHeight,
  //       popupWidth: elementDOM.offsetWidth,
  //     };
  //     return positionalStyle;
  //   };

  //   const findConditionals = node => ({
  //     ...(((node.type === 'conditional') && (node.data.get('whenFalse') === '') ) ? {
  //       [node.key]: {
  //         id: node.data.get('id'),
  //         whenTrue: node.data.get('whenTrue'),
  //         whenFalse: node.data.get('whenFalse'),
  //         position: findPosition(node.key),
  //         currentText: node.text,
  //         isFalse: node.text === node.data.get('whenFalse'),
  //       }
  //     } : {}),
  //     ...(node.nodes || [])
  //       .map(findConditionals)
  //       .reduce((props, node) => ({ ...props, ...node }), {})
  //   });

  //   let listData = {};
  //   const findListVariables = (node) =>{
  //     if((node.type === 'ul_list') ||(node.type ==='ol_list' )){
  //       (node.nodes || []).forEach((listNode,index) => {
  //        if(listNode.getInlinesByType("variable").size > 0){
  //           listData[listNode.key] = {
  //               currentText: listNode.text,
  //               head: index === 0,
  //               tail: index === (node.nodes.size - 1),
  //               parentKey: node.key,
  //               position: findPosition(listNode.key)
  //             }
  //        }
  //       })
  //     }
  //     return {
  //       ...listData,
  //       ...(node.nodes || [])
  //       .map(findListVariables)
  //       .reduce((props, node) => ({ ...props, ...node }), {})
  //     }

  //   }

  //   const newState = findConditionals(props.clauseNode);
  //   const foundListVariables = findListVariables(props.clauseNode)

  //   setConditionals(newState);
  //   setListVariables(foundListVariables)
  // }, [props.clauseNode, props.editor]);

  // const toggleConditional = (key) => {
  //   const selectionNodeKey = key || props.editor.value.selection.focus.key - 1;
  //   const selectedConditional = conditionals[selectionNodeKey];

  //   if (selectedConditional) {
  //     const selectionTextNode = props.editor.value.document.getNode(key);

  //     const newInlineJSON = {
  //       object: 'inline',
  //       type: 'conditional',
  //       data: {
  //         id: selectedConditional.id,
  //         whenTrue: selectedConditional.whenTrue,
  //         whenFalse: selectedConditional.whenFalse
  //       },
  //       nodes: [
  //         {
  //           object: 'text',
  //           text: selectionTextNode.text === selectedConditional.whenTrue
  //             ? selectedConditional.whenFalse
  //             : selectedConditional.whenTrue,
  //           marks: []
  //         }
  //       ]
  //     };
  //     const newInlineSlate = Inline.fromJSON(newInlineJSON);
  //     props.editor.replaceNodeByKey(selectionNodeKey, newInlineSlate);
  //   }
  // };

  // const addList = (listLastkey,parentKey) =>{
  //   const listLastNode = props.editor.value.document.getNode(listLastkey);
  //   const parentNode = props.editor.value.document.getNode(parentKey);

  //   const newBlockJSON = regenerateKey(listLastNode);
  //   const newBlockSlate = Block.fromJSON(newBlockJSON);

  //   let newNode = parentNode.nodes.asMutable().push(newBlockSlate);

  //  const updatedList = {
  //       data: parentNode.data,
  //       key: parentNode.key,
  //       object: parentNode.object,
  //       text: parentNode.text,
  //       type: parentNode.type,
  //       nodes:newNode
  //  }

  //   props.editor.replaceNodeByKey(parentKey,Block.fromJSON(updatedList));

  // }

  // const removeList = (key) => props.editor.removeNodeByKey(key);

  const testIconProps = {
    'aria-label': testIcon.type,
    width: '19px',
    height: '19px',
    viewBox: '0 0 16 20',
    // clauseIconColor: clauseProps.CLAUSE_ICONS
  };

  const editIconProps = {
    'aria-label': editIcon.type,
    width: '19px',
    height: '19px',
    viewBox: '0 0 19 19',
    // clauseIconColor: clauseProps.CLAUSE_ICONS
  };

  const deleteIconProps = {
    'aria-label': deleteIcon.type,
    width: '19px',
    height: '19px',
    viewBox: '0 0 12 15',
    // clauseIconColor: clauseProps.CLAUSE_ICONS
  };

  // const conditionalIconProps = {
  //   currentHover: hovering,
  //   toggleConditional,
  // };
  // const ListIconProps = {
  //   currentHover: hovering,
  //   removeList,
  //   addList
  // };
  return (
    <S.ClauseWrapper
      src={props.templateUri}
      id={props.clauseId}
      className='clause'
      onMouseEnter={() => setHovering(true)}
      onMouseLeave={() => setHovering(false)}
    >
    {/* {
      !props.readOnly
      && Object.entries(conditionals).map(([key, value]) => (
        value.isFalse && (value.whenFalse === '')
          && <ConditionalBoolean
              key={key}
              conditionalStyle={value.position.popupStyle}
              slateKey={key}
              nodeValue={value}
              {...conditionalIconProps}
            />
      ))
    }
    {
       Object.entries(listVariables).map(([key, value]) => (
        <ClauseVariableList
              key={key}
              listIconStyle={value.position.popupStyle}
              slateListKey={key}
              nodeValue={value}
              {...ListIconProps}
            />
      ))
    } */}
      <S.ClauseBackground
        clauseborder={clauseProps.CLAUSE_BORDER}
        clausebg={clauseProps.CLAUSE_BACKGROUND}
        contentEditable={false}
      />

      <S.ClauseHeader
        currentHover={hovering}
        headerfont={clauseProps.HEADER_FONT}
        headercolor={clauseProps.HEADER_COLOR}
        headerbg={clauseProps.CLAUSE_BACKGROUND}
        contentEditable={false}
      >
        {(hoveringHeader && header.length > 54)
          && <S.HeaderToolTipWrapper>
            <S.HeaderToolTip>
              {title + clauseProps.HEADER_TITLE}
            </S.HeaderToolTip>
          </S.HeaderToolTipWrapper>
        }
        <S.HeaderToolTipText
          onMouseEnter={() => setHoveringHeader(true)}
          onMouseLeave={() => setHoveringHeader(false)}
        >
          {header}
        </S.HeaderToolTipText>
      </S.ClauseHeader>
      { !props.readOnly
        && <>
          <S.TestWrapper
            {...iconWrapperProps}
            onMouseEnter={() => setHoveringTestIcon(true)}
            onMouseLeave={() => setHoveringTestIcon(false)}
            onClick={() => clauseProps.CLAUSE_TEST_FUNCTION(props)}
          >
            <S.ClauseIcon
              {...testIconProps}
              hovering={hoveringTestIcon}
            >
              {testIcon.icon()}
            </ S.ClauseIcon>
            {(hoveringTestIcon)
              && <S.HeaderToolTipWrapper>
                <S.HeaderToolTip>
                  Test
                </S.HeaderToolTip>
              </S.HeaderToolTipWrapper>
            }
          </S.TestWrapper>
          <S.EditWrapper
            {...iconWrapperProps}
            onMouseEnter={() => setHoveringEditIcon(true)}
            onMouseLeave={() => setHoveringEditIcon(false)}
            onClick={() => clauseProps.CLAUSE_EDIT_FUNCTION(props)}
          >
            <S.ClauseIcon
              {...editIconProps}
              hovering={hoveringEditIcon}
            >
              {editIcon.icon()}
            </ S.ClauseIcon>
            {(hoveringEditIcon)
              && <S.HeaderToolTipWrapper>
                <S.HeaderToolTip>
                  Edit
                </S.HeaderToolTip>
              </S.HeaderToolTipWrapper>
            }
          </S.EditWrapper>
          <S.DeleteWrapper
            {...iconWrapperProps}
            onMouseEnter={() => setHoveringDeleteIcon(true)}
            onMouseLeave={() => setHoveringDeleteIcon(false)}
            onClick={() => clauseProps.CLAUSE_DELETE_FUNCTION(props)}
          >
            <S.ClauseIcon
              {...deleteIconProps}
              hovering={hoveringDeleteIcon}
            >
              {deleteIcon.icon()}
            </ S.ClauseIcon>
            {(hoveringDeleteIcon)
              && <S.HeaderToolTipWrapper>
                <S.HeaderToolTip>
                  Delete
                </S.HeaderToolTip>
              </S.HeaderToolTipWrapper>
            }
          </S.DeleteWrapper>
        </>
      }
      <S.ClauseBody
        bodyfont={clauseProps.BODY_FONT}
        variablecolor={clauseProps.VARIABLE_COLOR}
        conditionalcolor={clauseProps.CONDITIONAL_COLOR}
        computedcolor={clauseProps.COMPUTED_COLOR}
        {...props.attributes}
      >
        {props.children}
      </S.ClauseBody>
  </S.ClauseWrapper>
  );
}

ClauseComponent.propTypes = {
  attributes: PropTypes.PropTypes.shape({
    'data-key': PropTypes.string,
  }),
  children: PropTypes.arrayOf(PropTypes.object).isRequired,
  clauseId: PropTypes.string,
  clauseNode: PropTypes.any,
  clauseProps: PropTypes.shape({
    BODY_FONT: PropTypes.string,
    CLAUSE_BACKGROUND: PropTypes.string,
    CLAUSE_BORDER: PropTypes.string,
    CLAUSE_ICONS: PropTypes.string,
    CLAUSE_DELETE_FUNCTION: PropTypes.func,
    CLAUSE_EDIT_FUNCTION: PropTypes.func,
    CLAUSE_TEST_FUNCTION: PropTypes.func,
    COMPUTED_COLOR: PropTypes.string,
    HEADER_COLOR: PropTypes.string,
    HEADER_FONT: PropTypes.string,
    HEADER_TITLE: PropTypes.string,
    VARIABLE_COLOR: PropTypes.string,
    CONDITIONAL_COLOR: PropTypes.string,
  }),
  editor: PropTypes.any,
  errors: PropTypes.object,
  readOnly: PropTypes.bool,
  removeFromContract: PropTypes.func,
  templateUri: PropTypes.string.isRequired,
};

export default ClauseComponent;
