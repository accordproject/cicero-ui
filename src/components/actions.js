const titleStart = input => input.lastIndexOf('/');
const titleEnd = input => input.indexOf('@');
const titleReducer = input => input.slice((titleStart(input) + 1), titleEnd(input));
const titleSpacer = input => input.replace(/-/g, ' ');
const titleCaps = input => input.toUpperCase();

export const titleGenerator = (input) => {
  const reducedTitle = titleReducer(input);
  const spacedTitle = titleSpacer(reducedTitle);
  const finalTitle = titleCaps(spacedTitle);
  return finalTitle;
};

export const headerGenerator = (templateTitle, inputTitle) => {
  const title = titleGenerator(templateTitle);
  const header = inputTitle ? (title + inputTitle) : title;
  const truncatedTitle = ((header.length > 54) ? (`${header.slice(0, 54)}...`) : header);
  return truncatedTitle;
};
