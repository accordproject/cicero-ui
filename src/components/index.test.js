import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { headerGenerator, titleGenerator } from './actions';
import ClauseComponent from './ClauseComponent';

const templateTitle = 'ap://acceptance-of-delivery@0.12.1#721d1aa0999a5d278653e211ae2a64b75fdd8ca6fa1f34255533c942404c5c1f';
const clauseId = '721d1aa0999a5d278653e211ae2a64b75fdd8ca6fa1f34255533c942404c5c1f';

const props = {
  clauseProps: {},
  templateUri: templateTitle,
  clauseId,
  children: []
};

describe('<ClauseComponent />', () => {
  describe('on initialization', () => {
    it('renders component correctly', () => {
      const component = shallow(<ClauseComponent {...props} />);
      const tree = toJson(component);
      expect(tree).toMatchSnapshot();
    });
  });
});

describe('headerGenerator', () => {
  it('should truncate a title over 54 characters', () => {
    const inputTitle = ' - CLAUSE TEMPLATE H A B C D E F G H';
    const truncatedTitle = 'ACCEPTANCE OF DELIVERY - CLAUSE TEMPLATE H A B C D E F...';
    expect(headerGenerator(templateTitle, inputTitle)).toEqual(truncatedTitle);
  });

  it('should not truncate a title under 54 characters', () => {
    const inputTitle = ' - CLAUSE TEMPLATE';
    const truncatedTitle = 'ACCEPTANCE OF DELIVERY - CLAUSE TEMPLATE';
    expect(headerGenerator(templateTitle, inputTitle)).toEqual(truncatedTitle);
  });

  it('should handle missing `inputTitle`', () => {
    const truncatedTitle = 'ACCEPTANCE OF DELIVERY';
    expect(headerGenerator(templateTitle)).toEqual(truncatedTitle);
  });
});

describe('titleGenerator',() => {
  it('should generate a title', () => {
    expect(titleGenerator(templateTitle)).toEqual('ACCEPTANCE OF DELIVERY');
  })
})