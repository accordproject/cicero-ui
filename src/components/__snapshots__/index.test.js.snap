// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClauseComponent /> on initialization renders page correctly 1`] = `
<styled.div
  id="721d1aa0999a5d278653e211ae2a64b75fdd8ca6fa1f34255533c942404c5c1f"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <styled.div />
  <styled.div
    currentHover={false}
  >
    ACCEPTANCE OF DELIVERYundefined
  </styled.div>
  <styled.div
    currentHover={false}
  >
    <styled.svg
      aria-label={[Function]}
      height="19px"
      onClick={[Function]}
      viewBox="0 0 12 15"
      width="15px"
    >
      <g
        fillRule="evenodd"
        id="Page-1"
        stroke="none"
        strokeWidth="1"
      >
        <g
          fillRule="nonzero"
          id="add/remove-files-expanded"
          transform="translate(-222.000000, -602.000000)"
        >
          <path
            d="M233.913531,604.599251 C234.037294,605.044944 234.082298,605 233.669757,604.996255 L222.328624,604.996255 C221.916083,604.996255 221.964838,605.044944 222.08485,604.599251 C222.178609,604.265918 222.223614,604.11236 222.223614,604.11236 C222.321123,603.764045 222.583649,603.730337 222.947436,603.670412 L224.942635,603.419476 C225.19016,603.374532 225.19016,603.374532 225.28767,603.164794 C225.613952,602.426966 225.711462,602 226.067748,602 L229.930633,602 C230.286919,602 230.388179,602.426966 230.714462,603.161049 C230.811971,603.374532 230.811971,603.374532 231.059496,603.41573 L233.050945,603.666667 C233.418482,603.726592 233.666007,603.726592 233.763516,604.074906 C233.763516,604.074906 233.823522,604.262172 233.913531,604.599251 Z M232.684263,606.808989 C233.306825,606.808989 233.336828,606.88764 233.299324,607.359551 L232.598004,616.445693 C232.537998,616.906367 232.492993,617 231.941688,617 L224.170912,617 C223.619607,617 223.574603,616.906367 223.514597,616.445693 L222.813277,607.359551 C222.775773,606.891386 222.805776,606.808989 223.428338,606.808989 L232.684263,606.808989 Z"
            id="trash"
          />
           
        </g>
      </g>
    </styled.svg>
  </styled.div>
  <styled.div />
</styled.div>
`;
